pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -829,9 +920,147 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
+			klog.Errorf("Unable to get resource configuration for pod %s", pod.Name)
+			return
+		}
+		setPodCgroupLimit := func(rName v1.ResourceName) error {
+			var err error
+			switch rName {
+			case v1.ResourceCPU:
+				err = pcm.SetPodCgroupCpuLimit(pod, podResources.CpuQuota, podResources.CpuPeriod, podResources.CpuShares)
+			case v1.ResourceMemory:
+				err = pcm.SetPodCgroupMemoryLimit(pod, *podResources.Memory)
+			}
+			if err != nil {
+				klog.Errorf("Failed to set %s cgroup limit for pod %s failed: %v", rName, pod.Name, err)
+			}
+			return err
+		}
+		// If resize results in net pod resource increase, set pod cgroup limits before resizing containers.
+		// If resize results in net pod resource decrease, set pod cgroup limits after resizing containers.
+		// If an error occurs at any point, abort. Let future syncpod iterations retry the unfinished stuff.
+		resizeContainers := func(rName v1.ResourceName, currPodCgLimit, newPodCgLimit int64) error {
+			var err error
+			if newPodCgLimit > currPodCgLimit {
+				if err = setPodCgroupLimit(rName); err != nil {
+					return err
+				}
+			}
+			if len(podContainerChanges.ContainersToUpdate[rName]) > 0 {
+				if err = m.updateContainerLimits(pod, podStatus, rName, podContainerChanges.ContainersToUpdate[rName]); err != nil {
+					klog.Errorf("updateContainerLimits for pod %q resource %s failed: %v", format.Pod(pod), rName, err)
+					return err
+				}
+			}
+			if newPodCgLimit < currPodCgLimit {
+				err = setPodCgroupLimit(rName)
+			}
+			return err
+		}
+		if len(podContainerChanges.ContainersToUpdate[v1.ResourceMemory]) > 0 || len(podContainerChanges.ContainersToRestart) > 0 {
+			currentPodMemoryLimit, err := pcm.GetPodCgroupMemoryLimit(pod)
+			if err != nil {
+				klog.Errorf("GetPodCgroupMemoryLimit for pod %s failed: %v", pod.Name, err)
+				return
+			}
+			if errResize := resizeContainers(v1.ResourceMemory, int64(currentPodMemoryLimit), *podResources.Memory); errResize != nil {
+				return
+			}
+		}
+		if len(podContainerChanges.ContainersToUpdate[v1.ResourceCPU]) > 0 || len(podContainerChanges.ContainersToRestart) > 0 {
+			currentPodCpuQuota, _, _, err := pcm.GetPodCgroupCpuLimit(pod)
+			if err != nil {
+				klog.Errorf("GetPodCgroupCpuLimit for pod %s failed: %v", pod.Name, err)
+				return
+			}
+			if errResize := resizeContainers(v1.ResourceCPU, currentPodCpuQuota, *podResources.CpuQuota); errResize != nil {
+				return
+			}
+		}
+		for _, idx := range podContainerChanges.ContainersToRestart {
+			start("container", containerStartSpec(&pod.Spec.Containers[idx]))
+		}
+	}
+
 	return
 }
 
+func (m *kubeGenericRuntimeManager) updateContainerLimits(pod *v1.Pod, podStatus *kubecontainer.PodStatus, resourceName v1.ResourceName, containersToUpdate []containerToUpdateInfo) error {
+	var updatedLimitsMap map[kubecontainer.ContainerID]struct {
+		cStatus *kubecontainer.ContainerStatus
+		rLimits v1.ResourceList
+	}
+	updatedLimitsMap = make(map[kubecontainer.ContainerID]struct {
```

Comment:
maybe define this struct in the function so we don't have to keep writing it out here and below

Reply:
Done in commit https://github.com/vinaykul/kubernetes/pull/1/commits/2d88cb142cb73073aa187edb2c8b6535ca416b8c